overlord/snapstate/snapmgr.go

```diff
@@ -1067,6 +1071,64 @@ func (m *SnapManager) localInstallCleanup() error {
 	return osutil.UnlinkManyAt(d, filenames)
 }
 
+func MockEnsuredMountsUpdated(m *SnapManager, ensured bool) (restore func()) {
+	osutil.MustBeTestBinary("ensured snap mounts can only be mocked from tests")
+	old := m.ensuredMountsUpdated
+	m.ensuredMountsUpdated = ensured
+	return func() {
+		m.ensuredMountsUpdated = old
+	}
+}
+
+func (m *SnapManager) ensureMountsUpdated() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	if m.ensuredMountsUpdated {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	allStates, err := All(m.state)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	var sysd systemd.Systemd
+	if len(allStates) > 0 {
+		if snapdenv.Preseeding() {
+			sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
+		} else {
+			sysd = systemd.New(systemd.SystemMode, nil)
+		}
+	}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+		squashfsPath := dirs.StripRootDir(info.MountFile())
+		whereDir := dirs.StripRootDir(info.MountDir())
+		if _, err = sysd.EnsureMountUnitFile(info.InstanceName(), info.Revision.String(), squashfsPath, whereDir, "squashfs"); err != nil {
+			return err
+		}
+	}
```

Comment:
I can extract this in a function. That can look better.

But in that extracted function, I do not see how it will be more obvious that sysd gets set when needed. Unless I missed something here.

Reply:
In the extracted function you can short-circuit the logic 
```
if len(allStates) == 0 {
   return nil
}

if snapdenv.Preseeding() {
    sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
} else {
    sysd = systemd.New(systemd.SystemMode, nil)
}
```

Now it's easy to observe that sysd is set in all cases.

Reply:
Thank you.

Reply:
I have changed the code so that `sysd` is never nil. But I only extracted the initialization of it.